File: product-service/src/main/java/com/github/buyalsky/productservice/service/ProductService.java
package com.github.buyalsky.productservice.service;

import com.github.buyalsky.productservice.dto.UpdateProductDto;
import com.github.buyalsky.productservice.entity.Product;

public interface ProductService {
    public Product getProductById(String id);
    public Product addProduct(Product product);
    public void deleteProduct(String productId);
    public Product updateProduct(String productId, UpdateProductDto updateProductDto);
    public Iterable<Product> getAllProducts();
}

File: shopping-cart-service/src/main/kotlin/com/github/buyalsky/shoppingcartservice/service/ProductService.kt
package com.github.buyalsky.shoppingcartservice.service

import com.github.buyalsky.shoppingcartservice.entity.ProductContainedInCarts
import reactor.core.publisher.Mono

interface ProductService {
    suspend fun findById(productId: String): ProductContainedInCarts
    suspend fun getCartsThatFollowProduct(productId: String): Set<String>
    suspend fun removeProduct(productId: String)
    suspend fun addFollowerToProduct(productId: String, userId: String): ProductContainedInCarts
}
